### PR TITLE
Update to v23.1.0-IOFreeze

### LINKED ISSUE #4789: Update to EnergyPlus 23.1.0
Enhancement Request
--------------
Once v23.1-IOFreeze is released, we should update to it.

## Detailed Description

https://github.com/NREL/EnergyPlus/blob/develop/src/Transition/InputRulesFiles/Rules22-2-0-to-23-1-0.md

### New additions

- [ ] `ZoneAirHeatBalanceAlgorithm`: new "Do Space Heat Balance for Sizing" and "Do Space Heat Balance for Simulation" fields. 
- [ ] `HeatPump:AirToWater:FuelFired:Heating` and `HeatPump:AirToWater:FuelFired:Cooling`: new objects. https://github.com/NREL/OpenStudio/pull/4806
- [ ] `OutputControl:Timestamp`: new object. https://github.com/NREL/OpenStudio/pull/4808
- [ ] `MaterialProperty:VariableAbsorptance`: new object. https://github.com/NREL/OpenStudio/issues/4801
- [ ] `SolarCollectorPerformance:PhotovoltaicThermal:BIPVT`: new object. https://github.com/NREL/OpenStudio/issues/4802

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

 - Fixes #4789
 - Also:
   - Add new FT/RT tests for `ZoneAirHeatBalanceAlgorithm` (and uncomment RT)

### Pull Request Author

 - [ ] Model API Changes / Additions
 - [ ] Any new or modified fields have been implemented in the EnergyPlus ForwardTranslator (and ReverseTranslator as appropriate)
 - [ ] Model API methods are tested (in `src/model/test`)
 - [ ] EnergyPlus ForwardTranslator Tests (in `src/energyplus/Test`)
 - [ ] If a new object or method, added a test in NREL/OpenStudio-resources: Add Link
 - [ ] If needed, added VersionTranslation rules for the objects (`src/osversion/VersionTranslator.cpp`)
 - [ ] Verified that C# bindings built fine on Windows, partial classes used as needed, etc.
 - [ ] All new and existing tests passes
 - [ ] If methods have been deprecated, update rest of code to use the new methods

**Labels:**

 - [ ] If change to an IDD file, add the label `IDDChange`
 - [ ] If breaking existing API, add the label `APIChange`
 - [ ] If deemed ready, add label `Pull Request - Ready for CI` so that CI builds your PR

### Review Checklist

This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] Code Style, strip trailing whitespace, etc.
 - [ ] All related changes have been implemented: model changes, model tests, FT changes, FT tests, VersionTranslation, OS App
 - [ ] Labeling is ok
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified

### CODE REVIEW

File: resources/model/OpenStudio.idd [hunk 1]
```diff
@@ -691,19 +691,32 @@ OS:ZoneAirContaminantBalance,
...
+       \type choice
+       \key No
+       \key Yes
+       \default No
+  A4; \field Do Space Heat Balance for Simulation
+       \note If yes, space heat balance will be calculated and reported during simulation.
+       \type choice
+       \key No
+       \key Yes
+       \default No
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply resources/model/OpenStudio.idd:719]: Ctor does harcode them though...

File: src/model/ZoneAirHeatBalanceAlgorithm.cpp [hunk 2]
```diff
@@ -143,6 +227,8 @@ namespace model {
     OS_ASSERT(getImpl<detail::ZoneAirHeatBalanceAlgorithm_Impl>());
 
     setAlgorithm("ThirdOrderBackwardDifference");
+    setDoSpaceHeatBalanceforSizing("No");
+    setDoSpaceHeatBalanceforSimulation("No");
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/ZoneAirHeatBalanceAlgorithm.cpp:231]: This is also wrong, it should be a bool

File: src/energyplus/Test/ZoneAirHeatBalanceAlgorithm_GTest.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+using namespace openstudio::model;
+using namespace openstudio;
+
+TEST_F(EnergyPlusFixture, ForwardTranslator_ZoneAirHeatBalanceAlgorithm) {
+  Model model;
+  ZoneAirHeatBalanceAlgorithm zoneAHBA = model.getUniqueModelObject<ZoneAirHeatBalanceAlgorithm>();
+
+  EXPECT_TRUE(zoneAHBA.setAlgorithm("AnalyticalSolution"));
+  EXPECT_TRUE(zoneAHBA.setDoSpaceHeatBalanceforSizing("Yes"));
+  EXPECT_TRUE(zoneAHBA.setDoSpaceHeatBalanceforSimulation("Yes"));
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/Test/ZoneAirHeatBalanceAlgorithm_GTest.cpp:60]: Should be a bool too

File: src/model/ZoneAirHeatBalanceAlgorithm.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/model/ZoneAirHeatBalanceAlgorithm.cpp]: Done in 95effb69a0be13f3c715f066bd9399bf8928ea40

File: CMakeLists.txt [hunk 4]
```diff
@@ -604,10 +603,10 @@ endif()
 if(UNIX)
   if(APPLE)
     if (ARCH MATCHES "arm64")
-      set(ENERGYPLUS_EXPECTED_HASH 35ba21891e7059987391f1ab59958025)
+      set(ENERGYPLUS_EXPECTED_HASH ee54b05e0da7fe77c098b8e28e2d5813)
```
[the remark this answers is not in the document]
Comment (jmarrec) [reply CMakeLists.txt]: Done in b2bcc51c0c3371eca3a8fd0292a8909cf3e5264c

File: src/energyplus/Test/ZoneAirHeatBalanceAlgorithm_GTest.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply src/energyplus/Test/ZoneAirHeatBalanceAlgorithm_GTest.cpp]: 95effb69a0be13f3c715f066bd9399bf8928ea40

File: resources/model/OpenStudio.idd (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmarrec) [reply resources/model/OpenStudio.idd:719]: Removed hardcoding in f15cb148a154e2045998cf7df87bfd2253d3a58e

**Comment by jmarrec** [comment]
@joseph-robertson Thanks for the PR. I've made the changes I wanted to see. Assuming CI passes and you're fine with the changes, this can merge right away!

File: resources/energyplus/ProposedEnergy+.idd [hunk 5]
```diff
@@ -1,4 +1,4 @@
-!IDD_Version 22.2.0
+!IDD_Version 23.1.0
```
Comment (jmarrec) [inline resources/energyplus/ProposedEnergy+.idd:1]: Side Note, but I'd find it clearer if you added all new objects to ProposedEnergy+.idd when you do this initial PR. The feature ones can build up upon it.

Comment (joseph-robertson) [reply resources/energyplus/ProposedEnergy+.idd:1]: Should with add this item to the https://github.com/NREL/OpenStudio/wiki/EnergyPlus-Version-Change wiki?

Comment (jmarrec) [reply resources/energyplus/ProposedEnergy+.idd:1]: Good idea, I added a clarification in the first bullet point that did mention "copying over changes".

**Pull request merged**